File: src/jsMain/js/utils/colorUtils.js
import chroma from 'chroma-js';

export function xy2polar(x, y) {
  let r = Math.sqrt(x * x + y * y);
  let phi = Math.atan2(y, x);
  return [r, phi];
}

export function polar2xy(r, phi) {
  const x = Math.cos(phi) * r;
  const y = Math.sin(phi) * r;
  return [x, y];
}

// rad in [-π, π] range
// return degree in [0, 360] range
export function rad2deg(rad) {
  return ((rad + Math.PI) / (2 * Math.PI)) * 360;
}
// degree in [0, 360] range
// return rad in [-π, π] range
export function deg2rad(deg) {
  return ((deg / 360) * (2 * Math.PI)) - Math.PI
}

export function xy2rgb(x, y, radius) {
  let [r, phi] = xy2polar(x, y);

  let deg = rad2deg(phi);
  let hue = deg;
  let saturation = r / radius;
  let value = 1.0;

  let [red, green, blue] = chroma.hsv(hue, saturation, value).rgb();
  return [red, green, blue];
}

export function rgb2xy(rgb, radius) {
  // Convert the color to polar coordinates
  const [hue, saturation] = chroma.rgb(...rgb).hsv();
  const deg = hue || 0; // hue will be NaN for white :-/
  const phi = deg2rad(deg);
  const r = saturation * radius;
  const [x, y] = polar2xy(r, phi);
  return [x + radius, y + radius];
}

File: src/jsMain/kotlin/baaahs/app/ui/gadgets/color/ColorUtils.kt
package baaahs.app.ui.gadgets.color

import baaahs.Color
import baaahs.geom.Vector2F
import kotlin.math.*

data class Polar(
    val r: Float,
    val phi: Float
) {
    fun toXy(): Vector2F {
        val x = cos(phi) * r
        val y = sin(phi) * r
        return Vector2F(x, y)
    }
}

fun Vector2F.toPolar(): Polar {
    val r = sqrt(x * x + y * y)
    val phi = atan2(y, x)
    return Polar(r, phi)
}


// rad in [-π, π] range
// return degree in [0, 1] range
fun rad2deg(rad: Float): Float {
    return ((rad + PI) / (2 * PI)).toFloat()
}

// degree in [0, 1] range
// return rad in [-π, π] range
fun deg2rad(deg: Float): Float {
    return ((deg * (2 * PI)) - PI).toFloat()
}

fun Polar.toColor(radius: Float, alpha: Float = 1f): Color {
    return Color.HSB(
        hue = rad2deg(phi),
        saturation = r / radius,
        brightness = 1.0f
    ).toRGB(alpha)
}

fun Color.toXy(radius: Float): Vector2F {
    // Convert the color to polar coordinates
    val hsb = toHSB()
    val deg = if (hsb.hue.isNaN()) 0f else hsb.hue; // hue will be NaN for white :-/
    val phi = deg2rad(deg)
    val r = hsb.saturation * radius
    return Polar(r, phi).toXy()
}
